src/common_labsoft/components/http_client.clj

```diff
@@ -0,0 +1,83 @@
+  (let [service-host (keyword (namespace key))
+        endpoint (keyword (name key))]
+    (render-route hosts service-host endpoint replace-map)))
+
+(defn resolve-token
+  "Receives a token. If its valid, returns it. If its not, generate a valid token and return it."
+  [token-component token service-name service-password]
+  (let [auth-payload {:auth/service-name service-name :auth/auth-password service-password}
+        config (:config token-component)
+        services (protocols.config/get! config :services)]
+    (if (protocols.token/verify token-component token)
+      token
+      (-> (render-route services :auth :service-token)
+          (client/post {:form-params auth-payload :content-type :json})
+          (:body)
+          (cheshire/parse-string true)
+          (:token/jwt)))))
+
+;; Transform methods: receives and returns request's data.
+(defn transform-url
+  "Generates a url if the data was provided with host/endpoint or keyworded url. Applies replace-map"
+  [data hosts]
+  (let [service (:host data)
+        endpoint (:endpoint data)
+        replace-map (:replace-map data {})
+        url (:url data)]
+    (if (keyword? url)
+      (let [url (render-keyworded-url hosts url (:replace-map data {}))]
+        (merge (dissoc data :replace-map) {:url url}))
+      (if (and (some? service) (some? endpoint))
+        (merge (dissoc data :replace-map :endpoint :host) {:url (render-route hosts service endpoint replace-map)})
+        data))))
+
+(defn transform-method
+  "Transforms request's method"
+  [data method]
+  (merge data {:method method}))
+
+(defrecord HttpClient
+  [config token]
+  protocols.http-client/HttpClient
+  (get-hosts [this] (protocols.config/get! config :services))
+  (raw-req! [this data] (client/request data))
+  (authd-req! [this data transform]
+    (let [service-name (protocols.config/get! config :service-name)
+          service-password (protocols.config/get! config :service-password)
+          valid-token (resolve-token this token service-name service-password)]
+      (if (not= valid-token @token)
+        (swap! token (fn [t] (atom valid-token))))
+      (protocols.http-client/raw-req! this (transform (merge {:headers {:authorization @token}} data)))))
+  (authd-req! [this data] (protocols.http-client/authd-req! this data identity))
+  (authd-get! [this data] (protocols.http-client/authd-req! this data (fn [req] (transform-method req :get))))
+  (authd-post! [this data] (protocols.http-client/authd-req! this data (fn [req] (transform-method req :post))))
+  (authd-patch! [this data] (protocols.http-client/authd-req! this data (fn [req] (transform-method req :patch))))
+  (authd-delete! [this data] (protocols.http-client/authd-req! this data (fn [req] (transform-method req :delete))))
+  (authd-put! [this data] (protocols.http-client/authd-req! this data (fn [req] (transform-method req :put))))
+  )
+
+(defn new-http-client [config token]
+  (map->HttpClient {:config config :token token}))
```

Comment:
Here I would create a version with arity `[]`. The config is going to be injected by the components lib, and also I think we can create the `Atom` token when calling this functions (there is no reason to receive it from outside)